Libraries/Opc.Ua.Client/Session/SessionAsync.cs

```diff
@@ -1051,6 +1061,300 @@ List<ServiceResult> errors
+
+                List<ReferenceDescriptionCollection> resultForPass = new List<ReferenceDescriptionCollection>();
+                resultForPass.AddRange(result);
+
+                List<ServiceResult> errorsForPass = new List<ServiceResult>();
+                errorsForPass.AddRange(errors);
+
+                int passCount = 0;
+
+                do
+                {
+                    int badNoCPErrorsPerPass = 0;
+                    int badCPInvalidErrorsPerPass = 0;
+                    int otherErrorsPerPass = 0;
+                    uint maxNodesPerBrowse = OperationLimits.MaxNodesPerBrowse;
+
+                    if (executeDefensively && ServerMaxContinuationPointsPerBrowse > 0)
+                    {
+                        maxNodesPerBrowse = ServerMaxContinuationPointsPerBrowse < maxNodesPerBrowse ? ServerMaxContinuationPointsPerBrowse : maxNodesPerBrowse;
+                    }
+
+                    // split input into batches
+                    int batchCount = 0;
+
+                    List<NodeId> nodesToBrowseForNextPass = new List<NodeId>();
+                    List<ReferenceDescriptionCollection> referenceDescriptionsForNextPass = new List<ReferenceDescriptionCollection>();
+                    List<ServiceResult> errorsForNextPass = new List<ServiceResult>();
+
+                    // loop over the batches
+                    foreach (var nodesToBrowseBatch in ((List<NodeId>)nodesToBrowseForPass).Batch<NodeId, List<NodeId>>(maxNodesPerBrowse))
+                    {
+                        int batchOffset = batchCount * (int)maxNodesPerBrowse;
+
+                        (
+                            List<ReferenceDescriptionCollection> resultForBatch,
+                            IList<ServiceResult> errorsForBatch
+                        )
+                        =
+                        await BrowseWithBrowseNextAsync(
+                                view,
+                                nodesToBrowseBatch,
+                                maxResultsToReturn,
+                                browseDirection,
+                                referenceTypeId,
+                                includeSubtypes,
+                                nodeClassMask
+                            ).ConfigureAwait(false);
+
+                        for (int ii = 0; ii < nodesToBrowseBatch.Count; ii++)
+                        {
+                            if (errorsForBatch[ii].StatusCode == StatusCodes.BadNoContinuationPoints ||
+                                errorsForBatch[ii].StatusCode == StatusCodes.BadContinuationPointInvalid)
+                            {
+                                nodesToBrowseForNextPass.Add(nodesToBrowseForPass[batchOffset + ii]);
+                                referenceDescriptionsForNextPass.Add(resultForPass[batchOffset + ii]);
+                                errorsForNextPass.Add(errorsForPass[batchOffset + ii]);
+                            }
```

Comment:
bad++ instead of iterating again through the list below

Reply:
yes, my bad, I had this fixed in the ManagedBrowse, but not in the ManagedBrowseAsync. Then I deleted the body of ManagedBrowse and called ManagedBrowseAsync instead. 
Fixed.